### PR TITLE
Refactor before OO Interface

### PR DESCRIPTION
**Pull request by rgrinberg**
This refactor adds more complete support for LSP and separates:

* Client notifications
* Server notifications
* Client requests
* Server requests

I've added some hierarchical names this time. I'm still not sure which approach is better.

After this PR the OO server will come.

### CODE REVIEW

**Comment by rgrinberg** [comment]
For fun, I've added a pseudo LSP client in OO style. @c-cube could you have a quick look over `Client`? I'm doing some awkward things to make sure that `request` & `notify` can be overriden.

File: lsp/src/header.ml [hunk 1]
```diff
@@ -0,0 +1,91 @@
+open Import
+
+(* TODO rewrite all this using faraday/angstrom *)
```
Comment (c-cube) [inline lsp/src/header.ml]: you can take code from https://github.com/c-cube/jsonrpc2/blob/master/src/jsonrpc2_core.ml#L515 if you want. Parsing these  is pretty simple.

**Comment by c-cube** [comment]
@rgrinberg right now `request` and `notify` are not implemented, right? What kind of overriding were you thinking about, I'm not seeing anything related in `Client`.

File: lsp/src/header.ml (hunk 1 above)
Comment (rgrinberg) [reply lsp/src/header.ml]: Would you mind sending a PR?

**Comment by rgrinberg** [comment]
> right now request and notify are not implemented, right?

Yes they aren't implemented

> What kind of overriding were you thinking about, I'm not seeing anything related in Client.

Oh, I guess removed it already. Suppose I want to call `request` or `notify` from within one of my default implementations. To achieve this, I need a bit of boilerplate to pass `t` to the created object itself.

**Comment by c-cube** [comment]
> Oh, I guess removed it already. Suppose I want to call `request` or `notify` from within one of my default implementations. To achieve this, I need a bit of boilerplate to pass `t` to the created object itself.

could you just provide methods for that? so you'd `self#notify` or `self#request`.

**Comment by rgrinberg** [comment]
> could you just provide methods for that? so you'd self#notify or self#request.

I would prefer if it was not possible to override these methods.

**Comment by c-cube** [comment]
A bit of a hack, but you might be able to use `val` for the functions.

```ocaml
class foo = object(self)
  val f = (fun x -> x+1)
  method g = f 40
end;;
class bar = object(self)
  inherit foo
  method g = f 50
end;;
let b = new bar;;
b#g;;
```

**Comment by rgrinberg** [comment]
I got rid of the experimental client API for now. The rest of the refactoring is straight forward I believe.

**Pull request merged**